src/util/vecdeque.h

```diff
@@ -0,0 +1,313 @@
+    friend void swap(VecDeque& a, VecDeque& b) noexcept { a.swap(b); }
+
+    /** Move-assign a deque. */
+    VecDeque& operator=(VecDeque&& other) noexcept
+    {
+        swap(other);
+        return *this;
+    }
+
+    /** Copy-construct a deque. */
+    VecDeque(const VecDeque& other) { *this = other; }
+    /** Move-construct a deque. */
+    VecDeque(VecDeque&& other) noexcept { swap(other); }
+
+    /** Equality comparison between two deques (only compares size+contents, not capacity). */
+    bool friend operator==(const VecDeque& a, const VecDeque& b)
+    {
+        if (a.m_size != b.m_size) return false;
+        for (size_t i = 0; i < a.m_size; ++i) {
+            if (a[i] != b[i]) return false;
+        }
+        return true;
+    }
+
+    /** Comparison between two deques, implementing lexicographic ordering on the contents. */
+    std::strong_ordering friend operator<=>(const VecDeque& a, const VecDeque& b)
+    {
+        size_t pos_a{0}, pos_b{0};
+        while (pos_a < a.m_size && pos_b < b.m_size) {
+            auto cmp = a[pos_a++] <=> b[pos_b++];
+            if (cmp != 0) return cmp;
+        }
+        return a.m_size <=> b.m_size;
+    }
+
+    /** Increase the capacity to capacity. Capacity will not shrink. */
+    void reserve(size_t capacity)
+    {
+        if (capacity > m_capacity) Reallocate(capacity);
+    }
+
+    /** Make the capacity equal to the size. The contents does not change. */
+    void shrink_to_fit()
+    {
+        if (m_capacity > m_size) Reallocate(m_size);
+    }
+
+    /** Construct a new element at the end of the deque. */
+    template<typename... Args>
+    void emplace_back(Args&&... args)
+    {
+        if (m_size == m_capacity) Reallocate((m_size + 1) * 2);
+        std::construct_at(m_buffer + BufferIndex(m_size), std::forward<Args>(args)...);
+        ++m_size;
+    }
+
+    /** Move-construct a new element at the end of the deque. */
+    void push_back(T&& elem) { emplace_back(std::move(elem)); }
+
+    /** Copy-construct a new element at the end of the deque. */
+    void push_back(const T& elem) { emplace_back(elem); }
+
+    /** Construct a new element at the beginning of the deque. */
+    template<typename... Args>
+    void emplace_front(Args&&... args)
+    {
+        if (m_size == m_capacity) Reallocate((m_size + 1) * 2);
+        std::construct_at(m_buffer + BufferIndex(m_capacity - 1), std::forward<Args>(args)...);
+        if (m_offset == 0) m_offset = m_capacity;
+        --m_offset;
+        ++m_size;
+    }
+
+    /** Copy-construct a new element at the beginning of the deque. */
+    void push_front(const T& elem) { emplace_front(elem); }
+
+    /** Move-construct a new element at the beginning of the deque. */
+    void push_front(T&& elem) { emplace_front(std::move(elem)); }
+
+    /** Remove the first element of the deque. Requires !empty(). */
+    void pop_front()
+    {
+        Assume(m_size);
+        std::destroy_at<T>(m_buffer + m_offset);
```

Comment:
Since you removed `<T>` after `construct_at`:
```suggestion
        std::destroy_at(m_buffer + m_offset);
```